### PR TITLE
v2 MobileKeypad (the v2 equivalent of ProvidedKeypad)

### PR DESCRIPTION
**Pull request by handeyeco**
## Summary:
**UPDATE**: we decided to push forward on this. Doing a straight overhaul of how we handle this logic proved to be tricky, so we're going to switch to this strategy and incrementally refactor once the dust settles. This way we can release the v2 keypad behind a feature flag now and start iterating.

---

I'm not super in love with the current structure of the LegacyKeypad: the primary export is a ProvidedKeypad which attaches a bunch of methods onto a class component which is then called via refs by at least 3 other components. This seems like a major anti-pattern and non-Reacty.

However I thought it would be an interesting though experiment to see what the minimum amount of changes we could need to add to have the v2 Keypad in mobile, and this requires creating a new ProvidedKeypad (v2) with the same API as the v1 ProvidedKeypad.

Surprisingly it just kind of works. ~~There's an issue with the "magic contextual cursor" button~~, ~~I have to figure out how to make the dismiss button work~~, and there's some style issues; but overall it seems like a good backup plan.

https://github.com/Khan/perseus/assets/16308368/492d2b0a-c49c-4daf-93e5-c9cc80fb5a41

Research ticket LC-1067

### CODE REVIEW

File: packages/math-input/src/components/keypad/index.tsx [hunk 1]
```diff
@@ -168,6 +168,8 @@ const styles = StyleSheet.create({
         gridTemplateRows: "repeat(4, 1fr)",
         backgroundColor: "#DBDCDD",
         maxHeight: 200,
-        maxWidth: 300,
+        // STOPSHIP we need a maxWidth somewhere for desktop
+        // but can't have it in mobile
+        // maxWidth: 300,
```
Comment (jeanettehead) [inline packages/math-input/src/components/keypad/index.tsx:173]: Would a media query help with this stopship?

Comment (handeyeco) [reply packages/math-input/src/components/keypad/index.tsx:173]: I think this is a good idea, but several other things change based on the `isMobile` flag in APIOptions and I'd be worried that the media query would eventually get out of sync with `isMobile`. I also tested removing `max-width` in dev tools with Khanmigo and nothing changed, so I think we might not need this at all anymore. 🤔 I'll fiddle around with it, thanks for the suggestion; didn't even think of that.

File: packages/math-input/src/components/keypad/keypad.tsx [hunk 2]
```diff
@@ -0,0 +1,171 @@
+import Color from "@khanacademy/wonder-blocks-color";
```
Comment (handeyeco) [inline packages/math-input/src/components/keypad/keypad.tsx:1]: I don't think anything really changed here. I just wanted to keep `index.ts` focused on exporting the pieces.

File: packages/math-input/src/components/keypad/mobile-keypad.tsx [hunk 3]
```diff
@@ -0,0 +1,158 @@
...
+import ReactDOM from "react-dom";
+
+import Key from "../../data/keys";
+import {View} from "../../fake-react-native-web/index";
+import {Cursor, KeypadConfiguration, KeyHandler} from "../../types";
+
+import Keypad from "./index";
+
+/**
+ * This is the v2 equivalent of v1's ProvidedKeypad. It follows the same
```
Comment (handeyeco) [inline packages/math-input/src/components/keypad/mobile-keypad.tsx:13]: To allow us to hot-swap the v1 and v2 keypads, this wrapper needs to have the same API as `/keypad-legacy/provided-keypad.tsx`

File: packages/math-input/src/components/keypad/keypad.tsx (hunk 2 above)
Comment (jeremywiebe) [reply packages/math-input/src/components/keypad/keypad.tsx:1]: In the future, doing this in two PRs would make this easier to review (file moves and then actual code changes).  Your comment helps though! Thanks. :)

File: packages/math-input/src/components/keypad/mobile-keypad.tsx [hunk 4]
```diff
@@ -0,0 +1,158 @@
...
+ * external API so that it can be hot-swapped with the v1 keypad and
+ * is responsible for connecting the keypad with MathInput and the Renderer.
+ *
+ * Ideally this strategy of attaching methods on the class component for
+ * other components to call will be replaced props/callbacks since React
+ * doesn't support this type of code anymore (functional components
+ * can't have methods attached to them).
```
Comment (jeremywiebe) [inline packages/math-input/src/components/keypad/mobile-keypad.tsx:20]: To be fair, they kinda do, but it's through the `[useImperativeHandle](https://react.dev/reference/react/useImperativeHandle)` hook and a forwardRef.

File: packages/math-input/src/components/keypad/mobile-keypad.tsx [hunk 5]
```diff
@@ -0,0 +1,158 @@
...
+
+import Key from "../../data/keys";
+import {View} from "../../fake-react-native-web/index";
+import {Cursor, KeypadConfiguration, KeyHandler} from "../../types";
+
+import Keypad from "./index";
+
+/**
+ * This is the v2 equivalent of v1's ProvidedKeypad. It follows the same
+ * external API so that it can be hot-swapped with the v1 keypad and
```
Comment (jeremywiebe) [inline packages/math-input/src/components/keypad/mobile-keypad.tsx:14]: I wonder if it'd help to pull out an `interface` that defines this shared API and then make the components implement that API. That'd help enforce what you've documented here.

File: packages/math-input/src/components/keypad/mobile-keypad.tsx [hunk 6]
```diff
@@ -0,0 +1,158 @@
...
+    ) => {
+        this.setState({keypadConfig: configuration});
+
+        // HACK(charlie): In Perseus, triggering a focus causes the keypad to
+        // animate into view and re-configure. We'd like to provide the option
+        // to re-render the re-configured keypad before animating it into view,
+        // to avoid jank in the animation. As such, we support passing a
+        // callback into `configureKeypad`. However, implementing this properly
+        // would require middleware, etc., so we just hack it on with
+        // `setTimeout` for now.
```
Comment (jeremywiebe) [inline packages/math-input/src/components/keypad/mobile-keypad.tsx:63]: I think this reference to "middleware" is now outdated... as I think that refers to Redux.

File: packages/math-input/src/components/keypad/mobile-keypad.tsx [hunk 7]
```diff
@@ -0,0 +1,158 @@
...
+
+        return (
+            <View
+                style={containerStyle}
+                ref={(element) => {
+                    if (!this.hasMounted && element) {
+                        // Append the dispatch methods that we want to expose
+                        // externally to the returned React element.
+                        const elementWithDispatchMethods = {
+                            ...element,
```
Comment (jeremywiebe) [inline packages/math-input/src/components/keypad/mobile-keypad.tsx:107]: This made me pause.  We're spreading a DOM node into the object we're returning. This is even more unclear by the fact that the `onElementMounded` prop is typed with an `any` argument. 

So two things:

1) is it possible to not spread the DOM node into this object (keep the element as a pure API)?
2) can we add better typing to the `onElementMounted` prop?

File: packages/math-input/src/components/keypad/mobile-keypad.tsx [hunk 8]
```diff
@@ -0,0 +1,158 @@
...
+
+                        this.hasMounted = true;
+                        this.props.onElementMounted?.(
+                            elementWithDispatchMethods,
+                        );
+                    }
+                }}
+            >
+                <Keypad
+                    sendEvent={async () => {}}
```
Comment (jeremywiebe) [inline packages/math-input/src/components/keypad/mobile-keypad.tsx:124]: Can you leave a `// TODO(jeremy)` here?  This analytics has taken waaaay too long. 😭

**Review by jeremywiebe — APPROVED** [review]
Makes sense.  👍

File: packages/math-input/src/components/keypad/mobile-keypad.tsx (hunk 6 above)
Comment (handeyeco) [reply packages/math-input/src/components/keypad/mobile-keypad.tsx:67]: I commented this...comment... and added a ticket to circle back and investigate if we need this hack at all post-Redux: https://khanacademy.atlassian.net/browse/LC-1080

File: packages/math-input/src/components/keypad/mobile-keypad.tsx (hunk 7 above)
Comment (jeremywiebe) [reply packages/math-input/src/components/keypad/mobile-keypad.tsx:113]: Totally fine with that! Thanks for creating the ticket.

File: packages/math-input/src/types.ts [hunk 9]
```diff
@@ -85,3 +87,12 @@ export type ActiveNodesObj = {
 export type LayoutProps = {initialBounds: Bound};
 
 export type ClickKeyCallback = (key: Key) => void;
+
+export interface KeypadAPI {
```
Comment (jeremywiebe) [inline packages/math-input/src/types.ts:91]: Nice!

**Pull request merged**